### PR TITLE
Database

### PR DESCRIPTION
**Pull request by olegdaniv**
Added database

### CODE REVIEW

File: app/src/main/java/com/example/topmovies/database/dao/BaseDao.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.example.topmovies.database.dao
+
+import androidx.room.Dao
+import androidx.room.Insert
+import androidx.room.OnConflictStrategy
+
+@Dao
+interface BaseDao<T> {
+
+    @Insert(onConflict = OnConflictStrategy.IGNORE)
```
Comment (YuriiTsap) [inline app/src/main/java/com/example/topmovies/database/dao/BaseDao.kt:10]: Why have you chosen IGNORE strategy?

File: app/src/main/java/com/example/topmovies/adapter/MoviesAdapter.kt [hunk 2]
```diff
@@ -9,23 +9,24 @@ import androidx.core.content.ContextCompat
...
 import com.example.topmovies.unit.RANK_UP
 import com.example.topmovies.unit.REPLACE_AFTER
+import com.example.topmovies.utils.GlideApp
+
 
 class MoviesAdapter(
     private val onItemClickListener: (String) -> Unit,
     private val onFavoriteMovieClick: (String, Boolean) -> Unit
-) : ListAdapter<Movie, MoviesAdapter.MovieViewHolder>(MovieDiffCallBack()) {
+) : ListAdapter<MovieEntity, MoviesAdapter.MovieViewHolder>(MovieDiffCallBack()) {
```
Comment (YuriiTsap) [inline app/src/main/java/com/example/topmovies/adapter/MoviesAdapter.kt]: What's the reason for operating ***Entity models at the UI level/layer?

File: app/src/main/java/com/example/topmovies/database/MovieDatabase.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+abstract class MovieDatabase : RoomDatabase() {
+
+    abstract fun movieDao(): MoviesDao
+
+    abstract fun movieDetails(): MovieDetailsDao
+
+    companion object {
+        private var INSTANCE: MovieDatabase? = null
+
+        fun getInstance(context: Context): MovieDatabase? {
```
Comment (iDobro) [inline app/src/main/java/com/example/topmovies/database/MovieDatabase.kt]: I think that it's not necessary and koin can create singleton as well.

File: app/src/main/java/com/example/topmovies/fragment/MoviesFragment.kt [hunk 4]
```diff
@@ -64,7 +64,13 @@ class MoviesFragment : BaseFragment() {
...
     }
 
-    private fun showMovieList(movies: List<Movie>) = with(binding) {
+    private fun showMovieList(movies: List<MovieEntity>, screen: EnumScreen) = with(binding) {
+        when (screen) {
+            EnumScreen.MOVIES -> emptyView.text =
+                resources.getString(R.string.movies_massage_no_movies)
+            EnumScreen.FAVORITE -> emptyView.text =
+                resources.getString(R.string.favorite_movies_massage_no_movies)
+        }
```
Comment (YuriiTsap) [inline app/src/main/java/com/example/topmovies/fragment/MoviesFragment.kt]: That doesn't fit into a single responsibility approach.

File: app/src/main/java/com/example/topmovies/repository/MovieRepository.kt [hunk 5]
```diff
@@ -1,58 +1,99 @@
...
+import com.example.topmovies.database.dao.MovieDetailsDao
+import com.example.topmovies.database.dao.MoviesDao
+import com.example.topmovies.model.*
 import com.example.topmovies.retrofit.MoviesApi
 import retrofit2.Call
 import retrofit2.Callback
 import retrofit2.Response
+import java.util.concurrent.ExecutorService
+
+class MovieRepository constructor(
```
Comment (YuriiTsap) [inline app/src/main/java/com/example/topmovies/repository/MovieRepository.kt:12]: Please regroup the methods a bit. keep the close to each other based on their role, i.e. insert and update shouldn't be apart.

File: app/src/main/java/com/example/topmovies/viewmodel/MovieViewModel.kt [hunk 6]
```diff
@@ -1,61 +1,60 @@
...
-    private val _favoriteMovies = mutableListOf<Movie>()
-    private val favoriteMovies: MutableLiveData<List<Movie>> = MutableLiveData(_favoriteMovies)
+    private val movies = MutableLiveData<List<MovieEntity>>()
+    private val favoriteMovies = MutableLiveData<List<MovieEntity>>()
     private var _errorMessage: String? = null
     val errorMessage = MutableLiveData(_errorMessage)
+    private val handler = Handler(Looper.getMainLooper())
 
-    fun getMoviesList(screen: EnumScreen): LiveData<List<Movie>> {
+    fun getMoviesList(screen: EnumScreen): LiveData<List<MovieEntity>> {
```
Comment (YuriiTsap) [inline app/src/main/java/com/example/topmovies/viewmodel/MovieViewModel.kt]: MovieEntity?

File: app/src/main/java/com/example/topmovies/viewmodel/MovieViewModel.kt [hunk 7]
```diff
@@ -1,61 +1,60 @@
...
         }
     }
 
     fun getMovies() {
-        movies.value ?: resolveMovies()
+        repository.getMovies {
+            val movieEntities = it
+            when (movieEntities.isEmpty()) {
+                true -> resolveMovies()
+                false -> handler.post { movies.value = movieEntities }
```
Comment (YuriiTsap) [inline app/src/main/java/com/example/topmovies/viewmodel/MovieViewModel.kt]: don't use when for true-false statements. if-else would work here. Or you can try to do that by using `takeIf()`

File: app/src/main/java/com/example/topmovies/viewmodel/MovieViewModel.kt [hunk 8]
```diff
@@ -1,61 +1,60 @@
...
         return when (screen) {
             EnumScreen.MOVIES -> movies
             EnumScreen.FAVORITE -> favoriteMovies
         }
     }
 
     fun getMovies() {
-        movies.value ?: resolveMovies()
+        repository.getMovies {
+            val movieEntities = it
```
Comment (YuriiTsap) [inline app/src/main/java/com/example/topmovies/viewmodel/MovieViewModel.kt]: Why for you need that? You already have that in the lambda declaration.
`repository.getMovies { movieEntities ->`

File: app/src/main/java/com/example/topmovies/viewmodel/MovieViewModel.kt [hunk 9]
```diff
@@ -1,61 +1,60 @@
...
 
-    fun getMoviesList(screen: EnumScreen): LiveData<List<Movie>> {
+    fun getMoviesList(screen: EnumScreen): LiveData<List<MovieEntity>> {
         return when (screen) {
             EnumScreen.MOVIES -> movies
             EnumScreen.FAVORITE -> favoriteMovies
         }
     }
 
     fun getMovies() {
```
Comment (YuriiTsap) [inline app/src/main/java/com/example/topmovies/viewmodel/MovieViewModel.kt:29]: What if I don't want to show anything before all data is loaded?
If your DB is empty you might end up in an inconsistent state.
For favorite movies, you aren't taking the empty DB condition into consideration.

File: app/src/main/java/com/example/topmovies/view/CircleAvatarView.kt [hunk 10]
```diff
@@ -54,7 +54,7 @@ class CircleAvatarView @JvmOverloads constructor(
                 label = typedArray.getString(R.styleable.CircleAvatarView_text) ?: ""
                 imageBitmap = BitmapFactory.decodeResource(
                     resources, typedArray.getResourceId(
-                        R.styleable.CircleAvatarView_src, 0
+                        R.styleable.CircleAvatarView_src, R.drawable.def_profile_image
```
Comment (YuriiTsap) [inline app/src/main/java/com/example/topmovies/view/CircleAvatarView.kt:57]: That wasn't part of your assignment.  Even if you are doing non-related work, that should be a separate commit.

File: app/src/main/java/com/example/topmovies/viewmodel/MovieViewModel.kt [hunk 11]
```diff
@@ -64,30 +63,25 @@ class MovieViewModel constructor(
...
-            mutableMovies[movies.value!!.indexOf(movie)] = copy
-            movies.postValue(mutableMovies)
-            _favoriteMovies.add(copy)
-            favoritePref.edit().putString(copy.id, "").apply()
+        repository.updateMovie(id, true)
+        repository.getMovies {
+            handler.post { movies.value = it }
+        }
+        repository.getFavoriteMovie {
+            handler.post { favoriteMovies.value = it }
```
Comment (YuriiTsap) [inline app/src/main/java/com/example/topmovies/viewmodel/MovieViewModel.kt]: Are you aware that you might run into an inconsistent data scenario?
What will happen if `getFavouriteMovies()` will execute prior to `updateMovies()`?

File: app/src/main/java/com/example/topmovies/viewmodel/MovieViewModel.kt [hunk 12]
```diff
@@ -64,30 +63,25 @@ class MovieViewModel constructor(
...
-            mutableMovies[movies.value!!.indexOf(movie)] = copy
-            movies.postValue(mutableMovies)
-            _favoriteMovies.remove(movie)
-            favoritePref.edit().remove(copy.id).apply()
+        repository.updateMovie(id, false)
+        repository.getMovies {
+            handler.post { movies.value = it }
+        }
+        repository.getFavoriteMovie {
+            handler.post { favoriteMovies.value = it }
```
Comment (YuriiTsap) [inline app/src/main/java/com/example/topmovies/viewmodel/MovieViewModel.kt]: Same here.
From the logical end, why would you need to get everything from the database?
Why not update the same movie? Why should we re-fetch everything?

File: app/src/main/java/com/example/topmovies/database/MovieDatabase.kt (hunk 3 above)
Comment (olegdaniv) [reply app/src/main/java/com/example/topmovies/database/MovieDatabase.kt]: Thank you for your comment. I did my own instance because, I wasn't sure if Koin dependency injection do instance like synchronized.

File: app/src/main/java/com/example/topmovies/database/dao/BaseDao.kt (hunk 1 above)
Comment (olegdaniv) [reply app/src/main/java/com/example/topmovies/database/dao/BaseDao.kt:10]: I use OnConflictStrategy.IGNORE strategy because in this case, I don't need to replace all data fields. 
In this case, if I need to update part of a field, better use another query like:
 @Query("UPDATE tb_movies SET rank = :rank,rankUpDown = :rankUpDown  WHERE id = :id")
    abstract fun updateMovie(id: String, rank: String, rankUpDown: String)

File: app/src/main/java/com/example/topmovies/database/MovieDatabase.kt [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+import com.example.topmovies.models.MovieDetailsEntity
+import com.example.topmovies.models.MovieEntity
+
+@Database(entities = [MovieEntity::class, MovieDetailsEntity::class], version = 1)
+abstract class MovieDatabase : RoomDatabase() {
+
+    abstract fun movieDao(): MoviesDao
+
+    abstract fun movieDetails(): MovieDetailsDao
+
```
Comment (iDobro) [inline app/src/main/java/com/example/topmovies/database/MovieDatabase.kt]: Delete the extra line.

File: app/src/main/java/com/example/topmovies/domain/GetMoviesUseCase.kt [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+) {
+
+    operator fun invoke(onResult: (List<Movie>) -> Unit, onFailed: (String) -> Unit) {
+        executor.execute {
+            val result = repository.loadMovies().map { it.toMovie() }
+            result.takeIf { it.isNotEmpty() }?.let {
+                handler.post { onResult(result) }
+            }
+                ?: loadMoviesUseCase({ movieObject ->
+//                    upsertMoviesUseCase(movieObject, {}, {})
```
Comment (iDobro) [inline app/src/main/java/com/example/topmovies/domain/GetMoviesUseCase.kt]: Clean up the code

File: app/src/main/java/com/example/topmovies/repository/MovieRepository.kt [hunk 15]
```diff
@@ -1,56 +1,99 @@
...
+                        if (it.errorMessage.isNotEmpty()) {
+                            onError(it.errorMessage)
+                        } else {
+                            onSuccess(it)
+                        }
                     }
-                } else onError(response.code().toString())
+                } else {
+                    onError(response.code().toString())
+
```
Comment (iDobro) [inline app/src/main/java/com/example/topmovies/repository/MovieRepository.kt]: Delete the extra line

File: app/src/main/java/com/example/topmovies/domain/GetMoviesUseCase.kt [hunk 16]
```diff
@@ -0,0 +1,31 @@
+package com.example.topmovies.domain
+
+import android.os.Handler
+import com.example.topmovies.models.Movie
+import com.example.topmovies.repository.MovieRepository
+import java.util.concurrent.ExecutorService
+
+
```
Comment (iDobro) [inline app/src/main/java/com/example/topmovies/domain/GetMoviesUseCase.kt:10]: Delete the extra line

File: app/src/main/java/com/example/topmovies/TopMoviesApp.kt [hunk 17]
```diff
@@ -23,7 +23,14 @@ class TopMoviesApp : Application() {
...
             androidLogger(Level.DEBUG)
             androidContext(this@TopMoviesApp)
-            modules(appModule)
+            modules(
+                appModule,
+                repositoryModule,
+                databaseModule,
+                viewModelModule,
+                networkModule,
+                useCase
```
Comment (iDobro) [inline app/src/main/java/com/example/topmovies/TopMoviesApp.kt:32]: modul*

File: app/src/main/java/com/example/topmovies/domain/GetMoviesUseCase.kt [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+) : UseCase<None, Result<List<Movie>>>() {
+
+    override fun run(params: None): Data<Result<List<Movie>>> {
+        var movies = repository.getMoviesEntity().map { it.toMovie() }
+        var error = ""
+        movies.ifEmpty {
+            val newMovie = repository.loadNewMovies()
+            if (newMovie.error.isNotEmpty()) {
+                error = newMovie.error
+                newMovie
```
Comment (iDobro) [inline app/src/main/java/com/example/topmovies/domain/GetMoviesUseCase.kt:24]: Why is it here? Remove this.

File: app/src/main/java/com/example/topmovies/domain/GetMoviesUseCase.kt [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+            } else {
+                repository.upsertMoviesEntity(newMovie.data.map { it.toMovie() })
+                movies = newMovie.data.map { it.toMovie() }
+            }
+        }
+        val favoriteMovies = repository.getFavoriteMoviesEntity(true).map { it.toMovie() }
+        return Data(Result(movies, favoriteMovies), error)
+    }
+
+    data class Result<T>(val movies: T, val favorite: T)
```
Comment (iDobro) [inline app/src/main/java/com/example/topmovies/domain/GetMoviesUseCase.kt:34]: Use "pair" to return this result.

File: app/src/main/java/com/example/topmovies/domain/UpdateMovieUseCase.kt [hunk 20]
```diff
@@ -0,0 +1,24 @@
+package com.example.topmovies.domain
+
+import android.os.Handler
+import com.example.topmovies.domain.GetMoviesUseCase.Result
+import com.example.topmovies.domain.UpdateMovieUseCase.Params
+import com.example.topmovies.models.Movie
+import com.example.topmovies.repository.MovieRepository
+import java.util.concurrent.ExecutorService
+
+class UpdateMovieUseCase(
```
Comment (iDobro) [inline app/src/main/java/com/example/topmovies/domain/UpdateMovieUseCase.kt:10]: Find a better name, with a "Favorite" key world.
